cli/ops/compiler.rs

```diff
@@ -1,13 +1,59 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+use super::dispatch_json::{JsonOp, Value};
+use crate::op_error::OpError;
+use crate::ops::json_op;
+use crate::ops::JsonOpDispatcher;
 use crate::state::State;
 use deno_core::CoreIsolate;
+use deno_core::CoreIsolateState;
+use deno_core::ZeroCopyBuf;
+use std::sync::Arc;
+use std::sync::Mutex;
 
-pub fn init(i: &mut CoreIsolate, _s: &State) {
+pub fn init(
+  i: &mut CoreIsolate,
+  _s: &State,
+  response: Arc<Mutex<Option<String>>>,
+) {
   let custom_assets = std::collections::HashMap::new();
   // TODO(ry) use None.
   // TODO(bartlomieju): is this op even required?
   i.register_op(
     "op_fetch_asset",
     crate::op_fetch_asset::op_fetch_asset(custom_assets),
   );
+
+  i.register_op(
+    "op_compiler_respond",
+    json_op(compiler_op(response, op_compiler_respond)),
+  );
+}
+
+pub fn compiler_op<D>(
+  response: Arc<Mutex<Option<String>>>,
+  dispatcher: D,
+) -> impl JsonOpDispatcher
+where
+  D: Fn(
+    Arc<Mutex<Option<String>>>,
+    Value,
+    &mut [ZeroCopyBuf],
+  ) -> Result<JsonOp, OpError>,
+{
+  move |_isolate_state: &mut CoreIsolateState,
+        args: Value,
+        zero_copy: &mut [ZeroCopyBuf]|
+        -> Result<JsonOp, OpError> {
+    dispatcher(response.clone(), args, zero_copy)
+  }
+}
+
+fn op_compiler_respond(
+  response: Arc<Mutex<Option<String>>>,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let mut r = response.lock().unwrap();
+  *r = Some(args.to_string());
```

Comment:
What's happening here? `response` is coming from `init()` ? Since `init()` is only called once per process... does that mean that the response is only possible to be set once? What if this is called twice?

Reply:
> What's happening here? `response` is coming from `init()` ? Since `init()` is only called once per process... 

That's just op registration, this op is used after TS compiler does it's job and returns the result. Before this change it was happening using `globalThis.postMessage()` using Web Worker APIs.

> does that mean that the response is only possible to be set once? What if this is called twice?

That's correct, and that has been the behavior for quite some time. TS compiler is a single-use thing, ie. only single request can be sent and then it must be discarded.

Reply:
Maybe you should add `assert!(r.is_none());` before this

Reply:
Done